Sources/NIOCore/ByteBuffer-hexdump.swift

```diff
@@ -0,0 +1,263 @@
+    /// and the last `maxBytes/2` of that, replacing the rest with `...`, i.e. `01 02 03 ... 09 11 12`.
+    ///
+    /// - parameters:
+    ///     - maxBytes: The maximum amount of bytes presented in the dump.
+    private func hexDumpPlain(maxBytes: Int) -> String {
+        // If the buffer lenght fits in the max bytes limit in the hex dump, just dump the whole thing.
+        if self.readableBytes <= maxBytes {
+            return self.hexDump(format: .plain)
+        }
+
+        var buffer = self
+
+        // Safe to force-unwrap because we just checked readableBytes is > maxBytes above.
+        let front = buffer.readSlice(length: maxBytes / 2)!
+        buffer.moveReaderIndex(to: buffer.writerIndex - maxBytes / 2)
+        let back = buffer.readSlice(length: buffer.readableBytes)!
+
+        let startHex = front.hexDumpPlain()
+        let endHex = back.hexDumpPlain()
+        return startHex + " ... " + endHex
+    }
+
+    /// Returns a String of detailed hex dump of this slice.
+    /// Intended to be used internally in ``hexDump(format:)``
+    /// - parameters:
+    ///     - lineOffset: an offset from the beginning of the outer buffer that is being dumped. It's used to print the line offset in hexdump -C format.
+    ///     - paddingBefore: the amount of space to pad before the first byte dumped on this line, used in center and right columns.
+    ///     - paddingAfter: the amount of sapce to pad after the last byte on this line, used in center and right columns.
+    private func _hexDumpLine(lineOffset: Int, paddingBefore: Int = 0, paddingAfter: Int = 0) -> String {
+        // Each line takes 78 visible characters + \n
+        var result = ""
+        result.reserveCapacity(79)
+
+        // Left column of the hex dump signifies the offset from the beginning of the dumped buffer
+        // and is separated from the next column with two spaces.
+        result += String(lineOffset, radix: 16, padding: 8)
+        result += "  "
+
+        // Center column consists of:
+        // - xxd-compatible dump of the first 8 bytes
+        // - space
+        // - xxd-compatible dump of the rest 8 bytes
+        // If there are not enough bytes to dump, the column is padded with space.
+
+        // If there's any padding on the left, apply that first.
+        result += String(repeating: " ", count: paddingBefore * 3)
+        // If the padding is 8 or more bytes, pad with extra space to match hexdump -C format.
+        if paddingBefore >= 8 {
+            result += " "
+        }
+
+        // Iterate over the bytes in the line, and dump them one by one, insert a separator if needed.
+        for (byteIndex, byte) in self.readableBytesView.enumerated() {
+            result += String(byte, radix: 16, padding: 2)
+            result += " "
+            if byteIndex + paddingBefore == 7 {
+                result += " "
+            }
+        }
```

Comment:
This part is a little hard to follow because of the extra spaces being added conditionally. Can I suggest something like this?

```swift
    result += String(repeating: " ", count: paddingBefore * 3)

    // Add the left side of the central column
    if paddingBefore < 8 {
        for byte in self.readableBytesView.prefix(8 - paddingBefore) {
            result += String(byte, radix: 16, padding: 2)
            result += " "
        }
    }

    // Add an extra space for the centre column.
    result += " "

    // Add the right side of the central column.
    let bytesToSkip = paddingBefore < 8 ? 8 - paddingBefore : 0
    for byte in self.readableBytesView.dropFirst(bytesToSkip) {
        result += String(byte, radix: 16, padding: 2)
        result += " "
    }
```

Reply:
Reworked it based on your feedback: https://github.com/apple/swift-nio/pull/2475/commits/d057109f9661b26df9cda8a9defb2618fd448a14 

That way, all `if` statements can go away, seems more readable?